File: sdks/php/src/Collection.php
<?php

declare(strict_types=1);

namespace Backlex;

/** A CRUD handle for one collection. Obtain via $client->from('slug'). */
final class Collection
{
    public function __construct(private Client $client, private string $slug)
    {
    }

    public function list(?array $query = null): array
    {
        return $this->client->request('GET', "/api/items/{$this->slug}" . Client::buildSearch($query));
    }

    /** Fluent builder that compiles to a ListQuery. */
    public function query(): QueryBuilder
    {
        return new QueryBuilder(fn ($q) => $this->list($q));
    }

    /** Single-function aggregate (count/sum/avg/min/max), optionally grouped. */
    public function aggregate(array $body): array
    {
        return $this->client->request('POST', "/api/items/{$this->slug}/aggregate", $body);
    }

    public function one(string $id): array
    {
        return $this->client->request('GET', "/api/items/{$this->slug}/{$id}");
    }

    public function create(array $data): array
    {
        return $this->client->request('POST', "/api/items/{$this->slug}", $data);
    }

    public function update(string $id, array $patch): array
    {
        return $this->client->request('PATCH', "/api/items/{$this->slug}/{$id}", $patch);
    }

    public function delete(string $id): array
    {
        return $this->client->request('DELETE', "/api/items/{$this->slug}/{$id}");
    }
}

File: sdks/dart/lib/src/collection.dart
part of backlex;

/// A CRUD handle for one collection. Obtain via `client.from('slug')`.
class Collection {
  final Client _client;
  final String _slug;

  Collection(this._client, this._slug);

  Future<dynamic> list([Map<String, dynamic>? query]) =>
      _client.request('GET', '/api/items/$_slug${Client.buildSearch(query)}');

  /// Fluent builder that compiles to a ListQuery.
  QueryBuilder query() => QueryBuilder((q) => list(q));

  /// Single-function aggregate (count/sum/avg/min/max), optionally grouped.
  /// `body` = `{'agg': 'sum', 'field': 'price', 'groupBy': 'status'}`.
  Future<dynamic> aggregate(Map<String, dynamic> body) =>
      _client.request('POST', '/api/items/$_slug/aggregate', body);

  /// [query] may carry expand/locale — the same params the list endpoint accepts.
  Future<dynamic> one(String id, [Map<String, dynamic>? query]) =>
      _client.request('GET', '/api/items/$_slug/$id${Client.buildSearch(query)}');

  Future<dynamic> create(Map<String, dynamic> data) =>
      _client.request('POST', '/api/items/$_slug', data);

  Future<dynamic> update(String id, Map<String, dynamic> patch) =>
      _client.request('PATCH', '/api/items/$_slug/$id', patch);

  Future<dynamic> delete(String id) => _client.request('DELETE', '/api/items/$_slug/$id');

  /// Flip a versioned item to published.
  Future<dynamic> publish(String id) =>
      _client.request('POST', '/api/items/$_slug/$id/publish');

  /// Flip a versioned item back to draft.
  Future<dynamic> unpublish(String id) =>
      _client.request('POST', '/api/items/$_slug/$id/publish?unpublish=1');
}
